staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -113,71 +114,130 @@ func WithPriorityAndFairness(
-
-					handler.ServeHTTP(w, innerReq)
-				}()
+				forgetWatch = fcIfc.RegisterWatch(requestInfo)
+
+				// Notify the main thread that we're ready to start the watch.
+				close(shouldStartWatchCh)
 
+				// Wait until the request is finished from the APF point of view
+				// (which is when its initialization is done).
 				watchInitializationSignal.Wait()
-			} else {
-				handler.ServeHTTP(w, innerReq)
 			}
-		}
 
-		// find the estimated "width" of the request
-		// TODO: Maybe just make it costEstimator and let it return additionalLatency too for the watch?
-		// TODO: Estimate cost should also take fcIfc.GetWatchCount(requestInfo) as a parameter.
-		width := widthEstimator.EstimateWidth(r)
-		digest := utilflowcontrol.RequestDigest{RequestInfo: requestInfo, User: user, Width: width}
+			// Ensure that an item can be put to resultCh asynchronously.
+			resultCh := make(chan interface{}, 1)
 
-		fcIfc.Handle(ctx, digest, note, func(inQueue bool) {
-			if inQueue {
-				noteWaitingDelta(1)
-			} else {
-				noteWaitingDelta(-1)
+			// Call Handle in a separate goroutine.
+			// The reason for it is that from APF point of view, the request processing
+			// finishes as soon as watch is initialized (which is generally orders of
+			// magnitude faster then the watch request itself). This means that Handle()
+			// call finishes much faster and for performance reasons we want to reduce
+			// the number of running goroutines - so we run the shorter thing in a
+			// dedicated goroutine and the actual watch handler in the main one.
+			go func() {
+				defer func() {
+					err := recover()
+					// do not wrap the sentinel ErrAbortHandler panic value
+					if err != nil && err != http.ErrAbortHandler {
+						// Same as stdlib http server code. Manually allocate stack
+						// trace buffer size to prevent excessively large logs
+						const size = 64 << 10
+						buf := make([]byte, size)
+						buf = buf[:runtime.Stack(buf, false)]
+						err = fmt.Sprintf("%v\n%s", err, buf)
+					}
+
+					// Ensure that the result is put into resultCh independently of the panic.
+					resultCh <- err
+				}()
+
+				// We create handleCtx with explicit cancelation function.
+				// The reason for it is that Handle() underneath may start additional goroutine
+				// that is blocked on context cancellation. However, from APF point of view,
+				// we don't want to wait until the whole watch request is processed (which is
+				// when it context is actually cancelled) - we want to unblock the goroutine as
+				// soon as the request is processed from the APF point of view.
+				//
+				// Note that we explicitly do NOT call the actuall handler using that context
+				// to avoid cancelling request too early.
+				handleCtx, handleCtxCancel := context.WithCancel(ctx)
+
+				fcIfc.Handle(handleCtx, digest, note, queueNote, execute)
+				handleCtxCancel()
+			}()
+
+			defer func() {
+				if err := recover(); err != nil {
```

Comment:
this is the primary goroutine, I don't think we need this `recover` and `panic` - we are not doing any cleanup here. The panic recover filter in the handler chain should suffice, i think.

Reply:
we need the defer func call but do we need the recover block?
```
				if err := recover(); err != nil {
					// In case of panic from inline handler, ensure that we clean up after
					// ourselves, in particular unblock hanging goroutines.
					defer panic(err)
				}
```

Reply:
OK - I was fairly convinced that deferred functions aren't called in cased of panic.
Which is clearly not true.

Fixed.